editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,627 @@
+	ERR_FAIL_INDEX_MSG(p_index, gradient->get_point_count(), "Gradient point is out of bounds.");
+
+	// Use pre_grab_offset to determine things for the undo/redo.
+	if (Math::is_equal_approx(pre_grab_offset, p_offset)) {
+		return;
+	}
+
+	int new_idx = predict_insertion_index(p_offset);
+
+	gradient->set_offset(p_index, pre_grab_offset); // Pretend the point started from its old place.
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Move Gradient Point"));
+	undo_redo->add_do_method(*gradient, "set_offset", pre_grab_index, p_offset);
+	undo_redo->add_do_method(this, "set_selected_index", new_idx);
+	undo_redo->add_undo_method(*gradient, "set_offset", new_idx, pre_grab_offset);
+	undo_redo->add_undo_method(this, "set_selected_index", pre_grab_index);
+	undo_redo->commit_action();
+	queue_redraw();
+}
+
+void GradientEdit::set_color(int p_index, const Color &p_color) {
+	ERR_FAIL_INDEX_MSG(p_index, gradient->get_point_count(), "Gradient point is out of bounds.");
+
+	Color old_color = gradient->get_color(p_index);
+	if (old_color == p_color) {
+		return;
+	}
+
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Recolor Gradient Point"), UndoRedo::MERGE_ENDS);
+	undo_redo->add_do_method(*gradient, "set_color", p_index, p_color);
+	undo_redo->add_undo_method(*gradient, "set_color", p_index, old_color);
+	undo_redo->commit_action();
+	queue_redraw();
+}
+
+void GradientEdit::_reverse_gradient() {
+	int new_selected_idx = (selected_index == -1) ? -1 : (gradient->get_point_count() - selected_index - 1);
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Reverse Gradient"), UndoRedo::MERGE_DISABLE, *gradient);
+	undo_redo->add_do_method(*gradient, "reverse");
+	undo_redo->add_do_method(this, "set_selected_index", new_selected_idx);
+	undo_redo->add_undo_method(*gradient, "reverse");
+	undo_redo->add_undo_method(this, "set_selected_index", selected_index);
+	undo_redo->commit_action();
+}
+
+void GradientEdit::set_selected_index(int p_index) {
+	selected_index = p_index;
+	queue_redraw();
+}
+
+void GradientEdit::set_snap_enabled(bool p_enabled) {
+	snap_enabled = p_enabled;
+	queue_redraw();
+	if (gradient.is_valid()) {
+		if (snap_enabled) {
+			gradient->set_meta(SNAME("_snap_enabled"), true);
+		} else {
+			gradient->remove_meta(SNAME("_snap_enabled"));
+		}
+	}
+}
+
+void GradientEdit::set_snap_count(int p_count) {
+	snap_count = p_count;
+	queue_redraw();
+	if (gradient.is_valid()) {
+		if (snap_count != GradientEditor::DEFAULT_SNAP) {
+			gradient->set_meta(SNAME("_snap_count"), snap_count);
+		} else {
+			gradient->remove_meta(SNAME("_snap_count"));
+		}
+	}
+}
+
+ColorPicker *GradientEdit::get_picker() {
+	return picker;
+}
+
+PopupPanel *GradientEdit::get_popup() {
+	return popup;
+}
+
+Size2 GradientEdit::get_minimum_size() const {
+	return Size2(0, 60) * EDSCALE;
```

Comment:
Is there a reason to override the method instead of setting custom minimum size in the constructor?

Reply:
Not sure, actually I didn't write this. A lot of it is just from merging gradient_editor.cpp and gradient_editor_plugin.cpp. I only tweaked the height a bit to accommodate the better-looking buttons. I mostly avoided touching parts of the code I don't entirely understand.

Reply:
If you suggest that overriding the method is inferior, maybe I could clean up declarations like these? I think I remember this being on the CurveEditor too, and it's probably in a lot more places.

Reply:
I don't think anything would break if you used custom minimum size. Can you give it a try?

Reply:
> If you suggest that overriding the method is inferior, maybe I could clean up declarations like these in a follow-up PR? I think I remember this being on the CurveEditor too, and it's probably in a lot more places.

I don't remember there being many such places, but if you're too concerned about the implications, this can be set aside for now.